File: wpp_polling_forwarder.py
import os
import time
import requests
import json
from datetime import datetime

# Configuração via variáveis de ambiente ou direto no script
WPP_HOST = os.getenv('WPP_HOST', '10.30.0.50')
WPP_PORT = int(os.getenv('WPP_PORT', '21465'))
WPP_TOKEN = os.getenv('WPP_TOKEN', '$2b$10$GHPF2x_zztSuaqZYviUsJulvGCJUFJXXLhI4TvdvxToFAGJzMEP02')
WPP_SESSION = os.getenv('WPP_SESSION', 'nova1')
N8N_WEBHOOK = os.getenv('N8N_WEBHOOK', 'https://n8n.fabioleal.com.br/webhook/whatsapp-agent')
POLL_INTERVAL = int(os.getenv('POLL_INTERVAL', '5'))  # segundos

# Cache de mensagens processadas
processed_ids = set()

def get_chats():
    url = f"http://{WPP_HOST}:{WPP_PORT}/api/{WPP_SESSION}/all-chats"
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {WPP_TOKEN}"
    }
    try:
        resp = requests.get(url, headers=headers, timeout=10)
        resp.raise_for_status()
        return resp.json()
    except Exception as e:
        print(f"❌ Erro ao buscar chats: {e}")
        return []

def forward_message(msg):
    payload = {
        "body": msg.get("body"),
        "sender": msg.get("from"),
        "timestamp": msg.get("t"),
        "isGroupMsg": msg.get("isGroupMsg", False),
        "chatId": msg.get("chatId"),
    }
    print(f"➡️ Payload para n8n: {json.dumps(payload, ensure_ascii=False)}")
    try:
        resp = requests.post(N8N_WEBHOOK, json=payload, timeout=10)
        print(f"✅ Enviado para n8n: {payload['sender']} - status {resp.status_code}")
        print(f"🔎 Resposta completa n8n: {resp.text}")
        if resp.status_code >= 400:
            print(f"⚠️ Erro HTTP ao enviar para n8n: {resp.status_code} - {resp.text}")
    except Exception as e:
        print(f"❌ Erro ao enviar para n8n: {e}")

def poll():
    print(f"🚀 Polling WPPConnect: http://{WPP_HOST}:{WPP_PORT}")
    print(f"🎯 Forwarding to n8n webhook: {N8N_WEBHOOK}")
    print(f"⏱️  Poll interval: {POLL_INTERVAL}s")
    while True:
        raw = get_chats()
        print(f"🔍 DEBUG chats type: {type(raw)} | value: {str(raw)[:200]}")
        chats = raw.get("response", []) if isinstance(raw, dict) else raw
        found = 0
        if not isinstance(chats, list):
            print("❌ Resposta inesperada da API, esperando lista.")
            time.sleep(POLL_INTERVAL)
            continue
        for chat in chats:
            if not isinstance(chat, dict):
                print(f"⚠️ Ignorando item não-dict: {chat}")
                continue
            print(f"🗂️ Chat: {json.dumps(chat, ensure_ascii=False)[:300]}")
            chat_id = chat.get("id", {}).get("_serialized")
            is_group = chat.get("id", {}).get("server") == "g.us"
            if not chat_id:
                print(f"❌ Chat sem id serializado: {chat}")
                continue
            if is_group:
                # Busca mensagens recentes do grupo
                url = f"http://{WPP_HOST}:{WPP_PORT}/api/{WPP_SESSION}/get-messages?chatId={chat_id}&isGroup=true&includeMe=true&limit=10"
                headers = {
                    "accept": "application/json",
                    "Authorization": f"Bearer {WPP_TOKEN}"
                }
                try:
                    resp = requests.get(url, headers=headers, timeout=10)
                    resp.raise_for_status()
                    msgs_raw = resp.json()
                    msgs = msgs_raw.get("response", []) if isinstance(msgs_raw, dict) else msgs_raw
                    print(f"📦 Mensagens recebidas de grupo {chat_id}: {len(msgs)}")
                    for msg in msgs:
                        msg_id = msg.get("id")
                        if msg_id and msg_id not in processed_ids:
                            print(f"📨 Nova mensagem: {msg.get('body')} de {msg.get('from')}")
                            forward_message(msg)
                            processed_ids.add(msg_id)
                            found += 1
                except Exception as e:
                    print(f"❌ Erro ao buscar mensagens de grupo {chat_id}: {e}")
                    # Fallback: tenta usar lastMessage se disponível
                    last_msg = chat.get("lastMessage")
                    if last_msg:
                        msg_id = last_msg.get("id")
                        print(f"🕵️ lastMessage grupo (fallback): {json.dumps(last_msg, ensure_ascii=False)}")
                        if msg_id and msg_id not in processed_ids:
                            print(f"📨 Nova mensagem grupo (fallback): {last_msg.get('body')} de {last_msg.get('from')}")
                            forward_message(last_msg)
                            processed_ids.add(msg_id)
                            found += 1
            else:
                # Para chats privados, usa lastMessage
                last_msg = chat.get("lastMessage")
                if not last_msg:
                    print(f"❌ Chat privado {chat_id} não tem lastMessage.")
                    continue
                msg_id = last_msg.get("id")
                print(f"🕵️ lastMessage privado: {json.dumps(last_msg, ensure_ascii=False)}")
                if msg_id and msg_id not in processed_ids:
                    print(f"📨 Nova mensagem privada: {last_msg.get('body')} de {last_msg.get('from')}")
                    forward_message(last_msg)
                    processed_ids.add(msg_id)
                    found += 1
        if found:
            print(f"🔔 {found} mensagem(ns) encaminhada(s)")
        time.sleep(POLL_INTERVAL)

if __name__ == "__main__":
    poll()

File: wpp-polling-forwarder.js
#!/usr/bin/env node

const https = require('https');
const http = require('http');

// Configurações
const WPP_HOST = process.env.WPP_HOST || '10.30.0.50';
const WPP_PORT = process.env.WPP_PORT || '21465';
const WPP_TOKEN = process.env.WPP_TOKEN || '';
const SESSION = process.env.WPP_SESSION || 'nova1';
const N8N_WEBHOOK = process.env.N8N_WEBHOOK || 'https://n8n.fabioleal.com.br/webhook/whatsapp-agent';
const POLL_INTERVAL = parseInt(process.env.POLL_INTERVAL || '5000'); // 5 segundos

// Armazena IDs de mensagens já processadas
const processedMessages = new Set();
const MAX_PROCESSED_CACHE = 1000;

console.log('🚀 WPPConnect Polling Forwarder started');
console.log(`📡 Polling WPPConnect: http://${WPP_HOST}:${WPP_PORT}`);
console.log(`🎯 Forwarding to n8n webhook: ${N8N_WEBHOOK}`);
console.log(`⏱️  Poll interval: ${POLL_INTERVAL}ms`);

// Função para fazer requisição HTTP
function makeRequest(options, data = null) {
    return new Promise((resolve, reject) => {
        const client = options.protocol === 'https:' ? https : http;
        const req = client.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve({ status: res.statusCode, data: JSON.parse(body) });
                } catch (e) {
                    resolve({ status: res.statusCode, data: body });
                }
            });
        });

        req.on('error', reject);
        if (data) req.write(JSON.stringify(data));
        req.end();
    });
}

// Função para buscar chats e suas últimas mensagens
async function getRecentMessages() {
    try {
        const options = {
            hostname: WPP_HOST,
            port: WPP_PORT,
            path: `/api/${SESSION}/all-chats`,
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${WPP_TOKEN}`,
                'accept': 'application/json'
            }
        };

        const response = await makeRequest(options);

        if (response.status === 200 && response.data.response) {
            const chats = response.data.response;

            // DEBUG: Imprime o primeiro chat para análise
            if (chats.length > 0) {
                const c = chats[0];
                console.log('🔍 DEBUG Chat structure:', JSON.stringify({
                    id: c.id,
                    lastMessage: c.lastMessage ? {
                        id: c.lastMessage.id,
                        t: c.lastMessage.t,
                        body: c.lastMessage.body
                    } : 'null',
                    unreadCount: c.unreadCount
                }));
            }

            const messages = [];

            // Pega mensagens recentes (últimos 30 segundos)
            const now = Date.now();
            const TIME_WINDOW = 30000; // 30 segundos

            for (const chat of chats) {
                if (chat.lastMessage) {
                    const msgTimestamp = (chat.lastMessage.t * 1000); // converte para ms se necessário

                    // Se a mensagem é recente (independente de lida ou não)
                    if (now - msgTimestamp < TIME_WINDOW) {
                        messages.push({
                            id: chat.lastMessage.id || `${chat.id}_${chat.lastMessage.t}`,
                            from: chat.id,
                            chatId: chat.id,
                            body: chat.lastMessage.body || chat.lastMessage.caption || '',
                            timestamp: msgTimestamp,
                            fromMe: chat.lastMessage.fromMe || false,
                            isGroupMsg: chat.isGroup || false,
                            unreadCount: chat.unreadCount || 0
                        });
                    }
                }
            }

            return messages;
        }
        return [];
    } catch (error) {
        console.error('❌ Error fetching chats:', error.message);
        return [];
    }
}

// Função para enviar mensagem para n8n
async function forwardToN8n(message) {
    try {
        const url = new URL(N8N_WEBHOOK);
        const webhookData = {
            event: 'onMessage',
            sender: message.from || message.chatId,
            body: message.body || message.content || '',
            from: message.from || message.chatId,
            isGroupMsg: message.isGroupMsg || false,
            timestamp: message.timestamp || message.t || Date.now(),
            messageId: message.id
        };

        const options = {
            hostname: url.hostname,
            port: url.protocol === 'https:' ? 443 : 80,
            path: url.pathname,
            method: 'POST',
            protocol: url.protocol,
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(JSON.stringify(webhookData))
            }
        };

        const response = await makeRequest(options, webhookData);

        if (response.status >= 200 && response.status < 300) {
            console.log(`✅ Message forwarded to n8n - ID: ${message.id}`);
            return true;
        } else {
            console.error(`❌ Failed to forward message - Status: ${response.status}`);
            return false;
        }
    } catch (error) {
        console.error('❌ Error forwarding to n8n:', error.message);
        return false;
    }
}

// Função principal de polling
async function poll() {
    try {
        const messages = await getRecentMessages();

        if (messages.length > 0) {
            console.log(`📨 Found ${messages.length} unread message(s)`);

            for (const message of messages) {
                // Verifica se já processamos esta mensagem
                if (processedMessages.has(message.id)) {
                    continue;
                }

                // Ignora mensagens enviadas por nós mesmos
                if (message.fromMe) {
                    processedMessages.add(message.id);
                    continue;
                }

                console.log(`📤 Processing message from ${message.from}: ${message.body?.substring(0, 50)}...`);

                const success = await forwardToN8n(message);

                if (success) {
                    processedMessages.add(message.id);

                    // Limita o tamanho do cache
                    if (processedMessages.size > MAX_PROCESSED_CACHE) {
                        const firstItem = processedMessages.values().next().value;
                        processedMessages.delete(firstItem);
                    }
                }
            }
        }
    } catch (error) {
        console.error('❌ Error in poll cycle:', error.message);
    }
}

// Inicia o polling
console.log('👂 Starting polling...\n');
setInterval(poll, POLL_INTERVAL);

// Primeira execução imediata
poll();

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down...');
    process.exit(0);
});

process.on('SIGTERM', () => {
    console.log('\n🛑 Shutting down...');
    process.exit(0);
});
